### PR TITLE
Ellipsoid fit for Mag Calibration

### PR DESCRIPTION
**Pull request by LorenzMeier**
@vilhjalmur89 This should go into the codebase after testing.

### CODE REVIEW

**Comment by jgoppert** [comment]
Should we switch it over to matrix as well?

**Comment by vilhjalmur89** [comment]
`make px4fmu-v2_default` fails with error:

`/home/pixhawk/gcc-arm-none-eabi-5_4-2016q2/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: firmware_nuttx section `.data' will not fit in region `flash'
/home/pixhawk/gcc-arm-none-eabi-5_4-2016q2/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: region flash' overflowed by 4508 bytes`

It does work with v4 though.

**Comment by bugobliterator** [comment]
@vilhjalmur89 px4fmu-v2 build uses only 1MB of flash due to issue in some of the older stm32f4s if you go beyond that much flash size, and this is the last drop. You can try disabling a module like uavcan to make it compile. Or if you are feeling lucky increase the size to 2M and try.

**Comment by vilhjalmur89** [comment]
@bugobliterator Alright, now it builds but I get `MAG 0 SELFTEST FAILED` in QGC after calibration.
Also xmag, ymag and zmag are not centered around 0, so I guess the offset is wrong then?

**Comment by bugobliterator** [comment]
@vilhjalmur89 what do you mean xmag, ymag and zmag not centred around zero? if you mean raw values xmag, ymag and zmag, they are not centred atleast not altogether, the Magnetometer vector value is somewhere in the range of 200-300 milligauss and is constant for particular region. Also is the compass you are calibrating external or internal, what is the result of the cal? Its possible that offsets generated are crossing the bar.

File: src/modules/commander/calibration_routines.cpp [hunk 1]
```diff
@@ -233,7 +234,299 @@ int sphere_fit_least_squares(const float x[], const float y[], const float z[],
...
+		_sphere_lambda /= lma_damping;
+		memcpy(fit1_params, fit2_params, sizeof(fit1_params));
+		fitness = fit2;
+
+	} else if (fit1 < _fitness) {
+		fitness = fit1;
+	}
+
+	//--------------------Levenberg-Marquardt-part-ends-here--------------------------------//
+	if (!isnan(fitness) && fitness < _fitness) {
```
Comment (LorenzMeier) [inline src/modules/commander/calibration_routines.cpp]: should be PX4_ISNAN() because `isnan` is not well-defined in a portable code sense.

**Comment by LorenzMeier** [comment]
@bkueng Can you fix up and bring in? You can disable uavcan on FMUv2 if you run out of flash.

File: src/modules/commander/mag_calibration.cpp [hunk 2]
```diff
@@ -568,34 +599,46 @@ calibrate_return mag_calibrate_all(orb_advert_t *mavlink_log_pub)
...
 
 
-	float sphere_x[max_mags];
-	float sphere_y[max_mags];
-	float sphere_z[max_mags];
-	float sphere_radius[max_mags];
+	float sphere_x[max_mags] = {0.0f};
+	float sphere_y[max_mags] = {0.0f};
+	float sphere_z[max_mags] = {0.0f};
+	float sphere_radius[max_mags] = {0.2f};
```
Comment (vilhjalmur89) [inline src/modules/commander/mag_calibration.cpp]: This only sets the first value, `sphere_radius = [0.2, 0.0, 0.0, ...]`
http://stackoverflow.com/questions/1065774/initialization-of-a-normal-array-with-one-default-value

**Comment by vilhjalmur89** [comment]
Seems to work now

**Comment by bkueng** [comment]
I managed to get a negative scale for the first test:
```
nsh> param show CAL_MAG0*
Symbols: x = used, + = saved, * = unsaved
x + CAL_MAG0_ID [56,77] : 263176
x   CAL_MAG0_ROT [57,78] : -1
x + CAL_MAG0_XOFF [58,79] : 0.2845
x + CAL_MAG0_XSCALE [59,80] : 1.2412
x + CAL_MAG0_YOFF [60,81] : 0.0740
x + CAL_MAG0_YSCALE [61,82] : 1.2585
x + CAL_MAG0_ZOFF [62,83] : 0.3864
x + CAL_MAG0_ZSCALE [63,84] : -0.1070
```

Second time was ok:
```
nsh> param show CAL_MAG0*
Symbols: x = used, + = saved, * = unsaved
x + CAL_MAG0_ID [56,77] : 263176
x   CAL_MAG0_ROT [57,78] : -1
x + CAL_MAG0_XOFF [58,79] : 0.2471
x + CAL_MAG0_XSCALE [59,80] : 1.0817
x + CAL_MAG0_YOFF [60,81] : 0.1359
x + CAL_MAG0_YSCALE [61,82] : 0.9963
x + CAL_MAG0_ZOFF [62,83] : 0.2513
x + CAL_MAG0_ZSCALE [63,84] : 0.9080
```

Maybe my motion was just not accurate enough or the algorithm is not robust. Either way I think it should tell you.

**Comment by bugobliterator** [comment]
@bkueng can you provide logs capturing the data in the first cal?

**Comment by bugobliterator** [comment]
@bkueng the shown result could be obtained if the samples collected less then half of the ellipsoid, and ending up minimizing at paraboloid(the negative scale suggests that). The way to fix it would be to add methods that ensure proper coverage. Also to confirm if that was the case I need the data.

**Comment by bkueng** [comment]
I didn't collect the data, but I can try to reproduce tomorrow and capture the sensor data.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

**Comment by LorenzMeier** [comment]
@bugobliterator Didn't you refactor the matrix math to use the matrix lib? Are commits missing here?

**Comment by mhkabir** [comment]
@LorenzMeier why aren't we using Matrix here?

**Comment by jgoppert** [comment]
@mhkabir @bugobliterator is working on it. I believe the current issue was that he needed a return code for a failed inversion when passing it a singular matrix. Matrix just returns a matrix, so maybe we would have to add another matrix inversion call that returns a status for that use case.

**Pull request merged**